2016/exercises/final_project/Alba-Petar-Rebecca-Simon/PolytopeProject/Gale_Diagram.cpp

```diff
@@ -0,0 +1,322 @@
+    
+}
+
+
+bool Gale_Diagram::isSimplex(int i, int j , int k, int l) const{
+    //Remove the columns i,j,k,l by selecting the other columns:
+    vector<int> pointIndices = findMissing(i,j,k,l);
+    /*cout << i << " " << j << " " << k << " " << l << endl;
+    for(int m=0; m<pointIndices.size(); m++){
+		cout << pointIndices[m] << " ";
+	}
+	cout << endl;*/
+    //Select points
+    vector<Point> fourPoints;
+    for(int m=0; m<pointIndices.size(); m++){
+        int pointIndex = pointIndices[m];
+        Point p(points[pointIndex].get_coordinates(), points[pointIndex].get_sign());
+        fourPoints.push_back(p);
+    }
+    
+    return check_intersecting(fourPoints);
+}
+void Gale_Diagram::writeGraph(string filename){
+    FILE * pFile;
+    
+    pFile = fopen((filename+".txt").c_str() , "w");
+    igraph_t graph = makeVertexFacetStructure();
+    igraph_write_graph_edgelist(&graph,pFile);
+    igraph_destroy(&graph);
+    fclose (pFile);
+}
+
+igraph_t Gale_Diagram::makeVertexFacetStructure() const{
+    int s = (int)points.size();
+    
+    igraph_t graph;
+    igraph_empty(&graph, 28, IGRAPH_UNDIRECTED);
+    
+    int facetCounter=s;
+    int counter = 0;
+    
+    for(int i=0; i<s-3; i++){
+        for(int j=i+1; j<s-2; j++){
+            for(int k=j+1; k<s-1; k++){
+                for(int l=k+1; l<s; l++){
+                    counter++;
+
+                    if(isSimplex(i, j, k, l)){
+                        igraph_vector_t edges;
+                        
+                        igraph_vector_init(&edges, 8);
+                        
+                        vector<int> pointIndicesSimplex = findMissing(i,j,k,l);
+                        for(int i = 0; i<4; i++){
+                            VECTOR(edges)[2*i] = facetCounter;
+                            VECTOR(edges)[2*i+1] = pointIndicesSimplex[i];
+                        }
+                        facetCounter++;
+                        
+                        igraph_add_edges(&graph, &edges, 0);
+                        
+                        //writeGraph(graph, ("counting"+to_string(counter)+".txt").c_str());
+                        
+                        igraph_vector_destroy(&edges);
+                    }
+                }
+            }
+        }
+    }
+    
+    igraph_to_undirected(&graph, IGRAPH_TO_UNDIRECTED_COLLAPSE, 0);
+    
+    
+    return graph;
+}
+
+bool Gale_Diagram::isIsomorphic(const Gale_Diagram &gale) const{
+    igraph_bool_t result = false;
+    igraph_t graph = makeVertexFacetStructure();
+    igraph_t otherGraph = gale.makeVertexFacetStructure();
+    igraph_isomorphic(&graph, &otherGraph, &result);
+    return result;
+}
+
+//For any three points, looks whether they are on a line.
+bool Gale_Diagram::isSimplicial() const{
+    int s = (int)points.size();
+    for(int i=0; i<s-3; i++){
+        for(int j=i+1; j<s-2; j++){
+            //If ij is a vertical line
```

Comment:
Use a function `next_combination` that you pull from the Internet, and you can reduce this entire function to two or three **readable** lines